cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,342 @@
+= CIP2016-12-16 - Constraints syntax
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Mats Rydberg <mats@neotechnology.com>
+
+[abstract]
+.Abstract
+--
+This CIP describes syntax and semantics for Cypher constraints.
+These are language constructs that impose restrictions on the shape of the data graph, and how statements are allowed to change it.
+--
+
+toc::[]
+
+== Background
+
+Cypher has a loose notion of a schema, in which nodes and relationships may take very heterogeneous forms, both in terms of properties and in graph patterns.
+Constraints allow us to mould the heterogeneous nature of the property graph into a more regular form.
+
+== Proposal
+
+This CIP specifies the general syntax for constraint definition (and constraint removal), and provides several examples of possible use cases for constraints.
+However, the specification does not otherwise specify or limit the space of expressible constraints that the syntax and semantics allow.
+
+=== Syntax
+
+The constraint syntax is defined as follows:
+
+.Grammar definition for constraint syntax.
+[source, ebnf]
+----
+constraint command    = create-constraint | drop-constraint ;
+create-constraint     = "CREATE", "CONSTRAINT", [ constraint-name ], "FOR", pattern, "REQUIRE", constraint-predicate, { "REQUIRE", constraint-predicate } ;
```

Comment:
I know there has been conversation about how `CREATE` is an overloaded word if used for constraints as well, since it is already used for creating nodes and edges. I just wanted to plant the idea of using the word `DEFINE` for constraints instead. i.e. `DEFINE CONSTRAINT FOR (p:Person) REQUIRE UNIQUE p.firstName, p.lastName`